In [ ]:
import pandas as pd
import math

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, accuracy_score

In [ ]:
x = [[3], [3], [1], [4], [1]]
y = [[4.3], [4.8], [2.5], [3.0], [4.0]]

y_hat = [[4.0], [3.5], [2.5], [2.5], [4.1]]

In [ ]:
mean_squared_error(y, y_hat)

0.4079999999999998

In [ ]:
mean_absolute_error(y, y_hat)

0.43999999999999984

In [ ]:
raiz_cuadrada = math.sqrt(mean_squared_error(y, y_hat))
raiz_cuadrada

0.6387487769068523

In [ ]:
mean_squared_error(y, y_hat) ** 0.5

0.6387487769068523

In [ ]:
y = [["NO PASA"], ["PASA"], ["PASA"], ["NO PASA"], ["PASA"]]
y_hat = [["NO PASA"], ["PASA"], ["PASA"], ["NO PASA"], ["NO PASA"]]
accuracy_score(y, y_hat)

0.8

In [ ]:
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.tree import plot_tree
from sklearn.model_selection import train_test_split

In [ ]:
model = DecisionTreeRegressor()

In [ ]:
train = pd.read_csv("/content/sample_data/california_housing_train.csv")

In [ ]:
train

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
2        -114.56     33.69                17.0        720.0           174.0   
3        -114.57     33.64                14.0       1501.0           337.0   
4        -114.57     33.57                20.0       1454.0           326.0   
...          ...       ...                 ...          ...             ...   
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   
16997    -124.30     41.84                17.0       2677.0           531.0   
16998    -124.30     41.80                19.0       2672.0           552.0   
16999    -124.35     40.54                52.0       1820.0           300.0   

       population  households  median_income  median_house_value  
0          1015.0       472.0         1.4936             66900.0  
1          1129.0       463.0         1.8200             80100.0  
2           333.0       117.0         1.6509             85700.0  
3           515.0       226.0         3.1917             73400.0  
4           624.0       262.0         1.9250             65500.0  
...           ...         ...            ...                 ...  
16995       907.0       369.0         2.3571            111400.0  
16996      1194.0       465.0         2.5179             79000.0  
16997      1244.0       456.0         3.0313            103600.0  
16998      1298.0       478.0         1.9797             85800.0  
16999       806.0       270.0         3.0147             94600.0  

[17000 rows x 9 columns]

In [ ]:
test = pd.read_csv("/content/sample_data/california_housing_test.csv")

In [ ]:
test

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0       -122.05     37.37                27.0       3885.0           661.0   
1       -118.30     34.26                43.0       1510.0           310.0   
2       -117.81     33.78                27.0       3589.0           507.0   
3       -118.36     33.82                28.0         67.0            15.0   
4       -119.67     36.33                19.0       1241.0           244.0   
...         ...       ...                 ...          ...             ...   
2995    -119.86     34.42                23.0       1450.0           642.0   
2996    -118.14     34.06                27.0       5257.0          1082.0   
2997    -119.70     36.30                10.0        956.0           201.0   
2998    -117.12     34.10                40.0         96.0            14.0   
2999    -119.63     34.42                42.0       1765.0           263.0   

      population  households  median_income  median_house_value  
0         1537.0       606.0         6.6085            344700.0  
1          809.0       277.0         3.5990            176500.0  
2         1484.0       495.0         5.7934            270500.0  
3           49.0        11.0         6.1359            330000.0  
4          850.0       237.0         2.9375             81700.0  
...          ...         ...            ...                 ...  
2995      1258.0       607.0         1.1790            225000.0  
2996      3496.0      1036.0         3.3906            237200.0  
2997       693.0       220.0         2.2895             62000.0  
2998        46.0        14.0         3.2708            162500.0  
2999       753.0       260.0         8.5608            500001.0  

[3000 rows x 9 columns]

In [ ]:
# Variables claves para determinar el costo de una casa
x_train = train[["longitude", "latitude", "housing_median_age", "median_income"]]
y_train = train["median_house_value"]

In [ ]:
x_test = test[["longitude", "latitude", "housing_median_age", "median_income"]]
y_test = test["median_house_value"]

In [ ]:
model.fit(x_train, y_train)

DecisionTreeRegressor()

In [ ]:
y_hat_train = model.predict(x_train)
y_hat_train

array([ 66900.,  80100.,  85700., ..., 103600.,  85800.,  94600.])

In [ ]:
y_hat_test = model.predict(x_test)
y_hat_test

array([500001., 176000., 228200., ...,  92700.,  83800., 500001.])

In [ ]:
# MSE - train
mean_squared_error(y_train, y_hat_train)

0.0

In [ ]:
# MSE - test
mean_squared_error(y_test, y_hat_test)

4249048618.1716666

In [ ]:
# MAE - train
mean_absolute_error(y_train, y_hat_train)

0.0

In [ ]:
# MAE - test
mean_absolute_error(y_test, y_hat_test)

40755.33833333333

In [ ]:
# RMSE - train
mean_squared_error(y_train, y_hat_train) ** 0.5

0.0

In [ ]:
# RMSE - test
mean_squared_error(y_test, y_hat_test) ** 0.5

65184.726878093665

In [ ]:
# Dividir los datos
# Aquí estoy utilizando una división simple en entrenamiento y prueba, pero podrías considerar validación cruzada.
x_train_split, x_val, y_train_split, y_val = train_test_split(x_train, y_train, test_size=0.2, random_state=42)

In [ ]:
# Entrenar y evaluar modelos con diferentes max_depth
best_model = None
best_val_score = float('inf')

for max_depth in range(1, 21):
    model = DecisionTreeRegressor(max_depth=max_depth, random_state=42)
    model.fit(x_train_split, y_train_split)

    train_pred = model.predict(x_train_split)
    val_pred = model.predict(x_val)

    train_score = mean_squared_error(y_train_split, train_pred)
    val_score = mean_squared_error(y_val, val_pred)

    print(f"max_depth={max_depth}, Train MSE: {train_score:.2f}, Validation MSE: {val_score:.2f}")

    if val_score < best_val_score:
        best_val_score = val_score
        best_model = model

max_depth=1, Train MSE: 9213173572.68, Validation MSE: 9286967569.19
max_depth=2, Train MSE: 7352333449.54, Validation MSE: 7447552472.34
max_depth=3, Train MSE: 6709161461.49, Validation MSE: 6770852264.20
max_depth=4, Train MSE: 5890059775.22, Validation MSE: 6053405404.27
max_depth=5, Train MSE: 5247392675.22, Validation MSE: 5599210971.24
max_depth=6, Train MSE: 4563461380.16, Validation MSE: 5016130753.24
max_depth=7, Train MSE: 3742053328.87, Validation MSE: 4384127303.69
max_depth=8, Train MSE: 3175962071.04, Validation MSE: 3947470005.33
max_depth=9, Train MSE: 2586274406.84, Validation MSE: 3719432067.57
max_depth=10, Train MSE: 2138497155.40, Validation MSE: 3785510744.91
max_depth=11, Train MSE: 1685585811.44, Validation MSE: 3681524055.07
max_depth=12, Train MSE: 1269459241.50, Validation MSE: 3760764177.08
max_depth=13, Train MSE: 930492815.31, Validation MSE: 3906077797.12
max_depth=14, Train MSE: 661381572.40, Validation MSE: 3956227028.12
max_depth=15, Train MSE: 443302

In [ ]:
# Evaluar el mejor modelo en el conjunto de prueba
test_pred = best_model.predict(x_test)
test_score = mean_squared_error(y_test, test_pred)
print(f"Best model: max_depth={best_model.max_depth}, Test MSE: {test_score:.2f}")

Best model: max_depth=11, Test MSE: 3650916628.05


Durante el proceso de entrenamiento y evaluación de modelos de árbol de decisión con distintas profundidades máximas (max_depth), se observó que el rendimiento en el conjunto de validación siguió un patrón general. A medida que aumentábamos la profundidad máxima, el error cuadrado medio (MSE) disminuía tanto en el conjunto de entrenamiento como en el conjunto de validación. Sin embargo, a partir de cierto punto, el rendimiento en el conjunto de validación comenzó a estabilizarse e incluso aumentar, lo que indica un posible sobreajuste.

El modelo que logró el mejor rendimiento en el conjunto de validación fue el que tenía una profundidad máxima de 11. Sin embargo, al evaluar este modelo en el conjunto de prueba independiente, se obtuvo un MSE de 3650916628.05. Esto sugiere que el modelo podría estar ligeramente sobreajustado, ya que su rendimiento en el conjunto de prueba es ligeramente peor que en el conjunto de validación.